core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -316,4 +464,58 @@ public boolean hasNext() {
             }
         }
     }
+
+    private final class ConcatenatedIteratorImpl implements Iterator<Attribute<?>> {
+
+        private final Iterator<Attribute<?>> thisIt;
+        private final Iterator<Attribute<?>> rootIt;
+
+        @Nullable
+        private Attribute<?> next;
+
+        ConcatenatedIteratorImpl(Iterator<Attribute<?>> thisIt, Iterator<Attribute<?>> rootIt) {
+            this.thisIt = thisIt;
+            this.rootIt = rootIt;
+
+            if (thisIt.hasNext()) {
+                next = thisIt.next();
+            } else if (rootIt.hasNext()) {
+                next = new CopyOnWriteAttribute<>(rootIt.next());
+            } else {
+                next = null;
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return next != null;
+        }
+
+        @Override
+        public Attribute<?> next() {
+            final Attribute<?> next = this.next;
+            if (next == null) {
+                throw new NoSuchElementException();
+            }
+
+            if (thisIt.hasNext()) {
+                this.next = thisIt.next();
+            } else {
+                // Skip the attribute in rootIt if it's in thisAttributeMap.
+                for (;;) {
+                    if (rootIt.hasNext()) {
+                        final Attribute<?> tempNext = rootIt.next();
+                        if (!hasOwnAttr(tempNext.key())) {
+                            this.next = new CopyOnWriteAttribute<>(tempNext);
+                            break;
+                        }
+                    } else {
+                        this.next = null;
+                        break;
```

Comment:
Need to throw a `NoSuchElementException`? Please add a test case.

Reply:
This is for prefetching. `NoSuchElementException ` is raised here https://github.com/line/armeria/pull/2322/files#diff-74634bdb2b0b3d623e3378b6a6fa9037R498

Reply:
Ahh, OK. Thanks!